### DISCUSSION TITLE
Fake Timers

### DISCUSSION THREAD

**Discussion by kudla** [avajs/ava#2855] [Q&A] [answered]
Hi there
Guys please what are the bets practices for mocking system time within AVA?

It had no luck to fake timers on `beforeEach` just on `before` hooks. but anyway it concurrently races mocking the time 

Searching for examples I've just found several suggestions to go with `serial` approach.  
But I hope it to be not the only solution as it nullifies one of the principal advantages  using AVA.

So for now I'm trying to setup test isolation over sandboxing subject modules with 
* [vm2](https://github.com/patriksimek/vm2)
* [sinon](https://sinonjs.org/releases/v1.17.6/fake-timers/)

but so far just stuck a little with vm2 it self.

So maybe there're some other flows exists which are more simple? 

Would you share your experience?
Any examples?

Thanks.

**Comment by novemberborn** [comment, marked as answer]
`setTimeout`, `Date`, etc are globals, so shared between all running code in the process or worker thread. You cannot fake them and then also run concurrent tests.
